packages/gatsby/src/schema/schema.js

```diff
@@ -1046,9 +1055,15 @@ const addImplicitConvenienceChildrenFields = ({
         )
       }
     }
-
-    typeComposer.addFields(createChildrenField(typeName))
-    typeComposer.addFields(createChildField(typeName))
+    // Set `@childOf` extension automatically
+    // This will cause convenience children fields like `childImageSharp`
+    // to be added in `addConvenienceChildrenFields` method.
+    // Also required for proper printing of the `@childOf` directive in the snapshot plugin
+    if (!childOfExtension?.types) {
+      childOfExtension = { types: [] }
+    }
+    childOfExtension.types.push(parentTypeName)
+    childTypeComposer.setExtension(`childOf`, childOfExtension)
```

Comment:
The essence of this PR: instead of adding fields directly we add the `childOf` directive here. And fields are always added in a single place now - where this directive is processed (in `addConvenienceChildrenFields`):

https://github.com/gatsbyjs/gatsby/blob/76a3b57455704f72778b41e01f00ed1b61cc6622/packages/gatsby/src/schema/schema.js#L972-L975